### PR TITLE
Precond setters

### PR DESCRIPTION
**Pull request by alanlamaswolox**
## Summary
Added preconditions to setters of Book and User models

## Trello
[link](https://trello.com/c/rBFpaLhD/29-agregar-validaci%C3%B3n-de-precondiciones-a-los-setters)

### CODE REVIEW

File: src/main/java/wolox/training/models/Book.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package wolox.training.models;
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java:2]: Complete the @Column annotation of the genre field.

In the set method of the same field add preconditions where it is verified that the field is not null or empty.

File: src/main/java/wolox/training/models/User.java [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package wolox.training.models;
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java:2]: Add an empty constructor to this class.

File: src/main/java/wolox/training/models/Book.java [hunk 3]
```diff
@@ -62,6 +63,7 @@ public String getIsbn() {
     }
 
     public void setIsbn(String isbn) {
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java:74]: Add the following validations:

- That the field is not empty

- That the field only contains numerical values

File: src/main/java/wolox/training/models/Book.java [hunk 4]
```diff
@@ -70,6 +72,7 @@ public int getPages() {
     }
 
     public void setPages(int pages) {
+        Preconditions.checkNotNull(pages, "pages must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: Add the following validations:

- That the field is not empty

- That the field value is greater than zero

File: src/main/java/wolox/training/models/Book.java [hunk 5]
```diff
@@ -78,6 +81,7 @@ public String getYear() {
     }
 
     public void setYear(String year) {
+        Preconditions.checkNotNull(year, "year must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: Add the following validations:

- That the field is not empty

- That the field only contains numerical values

File: src/main/java/wolox/training/models/Book.java [hunk 6]
```diff
@@ -86,6 +90,7 @@ public String getPublisher() {
     }
 
     public void setPublisher(String publisher) {
+        Preconditions.checkNotNull(publisher, "publisher must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: Add the following validations:

- That the field is not empty

File: src/main/java/wolox/training/models/Book.java [hunk 7]
```diff
@@ -94,6 +99,7 @@ public String getSubtitle() {
     }
 
     public void setSubtitle(String subtitle) {
+        Preconditions.checkNotNull(subtitle, "subtitle must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: Add the following validations:

- That the field is not empty

File: src/main/java/wolox/training/models/Book.java [hunk 8]
```diff
@@ -102,6 +108,7 @@ public String getAuthor() {
     }
 
     public void setAuthor(String author) {
+        Preconditions.checkNotNull(author, "author must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: Add the following validations:

- That the field is not empty

File: src/main/java/wolox/training/models/Book.java [hunk 9]
```diff
@@ -110,6 +117,7 @@ public String getTitle() {
     }
 
     public void setTitle(String title) {
+        Preconditions.checkNotNull(title, "title must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: Add the following validations:

- That the field is not empty

File: src/main/java/wolox/training/models/Book.java [hunk 10]
```diff
@@ -126,6 +134,7 @@ public String getImage() {
     }
 
     public void setImage(String image) {
+        Preconditions.checkNotNull(image, "image must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: Add the following validations:

- That the field is not empty

File: src/main/java/wolox/training/models/User.java [hunk 11]
```diff
@@ -56,6 +58,7 @@ public String getName() {
     }
 
     public void setName(String name) {
+        Preconditions.checkNotNull(name, "name must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: Add the following validations:

- That the field is not empty

File: src/main/java/wolox/training/models/User.java [hunk 12]
```diff
@@ -48,6 +49,7 @@ public String getUsername() {
     }
 
     public void setUsername(String username) {
+        Preconditions.checkNotNull(username, "username must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: Add the following validations:

- That the field is not empty

File: src/main/java/wolox/training/models/User.java [hunk 13]
```diff
@@ -64,6 +67,7 @@ public LocalDate getBirthDate() {
     }
 
     public void setBirthDate(LocalDate birthDate) {
+        Preconditions.checkNotNull(birthDate, "birthDate must be not null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: Add the following validations:

- That the field is not empty
- Validate if the birthday is earlier than the current date

File: src/main/java/wolox/training/models/Book.java [hunk 14]
```diff
@@ -1,5 +1,6 @@
 package wolox.training.models;
 
+import com.google.common.base.Preconditions;
 import io.swagger.annotations.Api;
 import io.swagger.annotations.ApiModel;
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java:4]: Add all error messages in a constant file.

File: src/main/java/wolox/training/models/User.java [hunk 15]
```diff
@@ -1,5 +1,6 @@
 package wolox.training.models;
 
+import com.google.common.base.Preconditions;
 import java.time.LocalDate;
 import java.util.ArrayList;
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java:5]: Add all error messages in a constant file.

**Review by Gianni-Nicolas — CHANGES_REQUESTED** [review]
Very good work. I leave you just a few comments 👍 😄

File: src/main/java/wolox/training/models/Book.java [hunk 16]
```diff
@@ -62,6 +63,7 @@ public String getIsbn() {
     }
 
     public void setIsbn(String isbn) {
+        Preconditions.checkNotNull(isbn, "isbn must be not null");
```
Comment (AleFalcon) [inline src/main/java/wolox/training/models/Book.java]: It's okay, that i would create a file with constants, and i would create a constant = "isbn must be not null"

File: src/main/java/wolox/training/models/Book.java (hunk 4 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/Book.java]: It's okay, that i would create a file with constants, and i would create a constant = "pages must be not null"

File: src/main/java/wolox/training/models/Book.java (hunk 5 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/Book.java]: Change constant

File: src/main/java/wolox/training/models/Book.java (hunk 6 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/Book.java]: Change constant

File: src/main/java/wolox/training/models/Book.java (hunk 7 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/Book.java]: Change constant

File: src/main/java/wolox/training/models/Book.java (hunk 8 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/Book.java]: Change constant

File: src/main/java/wolox/training/models/Book.java (hunk 9 above)
Comment (AleFalcon) [reply src/main/java/wolox/training/models/Book.java]: Change constant

File: src/main/java/wolox/training/models/Book.java (hunk 10 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/Book.java]: Change constant

File: src/main/java/wolox/training/models/User.java (hunk 12 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/User.java]: Change constant

File: src/main/java/wolox/training/models/User.java (hunk 11 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/User.java]: Change constant

File: src/main/java/wolox/training/models/User.java (hunk 13 above)
Comment (AleFalcon) [inline src/main/java/wolox/training/models/User.java]: Change constant

**Review by AleFalcon — COMMENTED** [review]
Any comments

File: src/main/java/wolox/training/models/Book.java (hunk 1 above)
Comment (Colum) [reply src/main/java/wolox/training/models/Book.java:2]: did you mean @Column ?

Comment (alanlamaswolox) [reply src/main/java/wolox/training/models/Book.java:2]: JAJAJAJAJAJAJAJAJAJAJAJAJAJA

File: src/main/java/wolox/training/models/Book.java [hunk 17]
```diff
@@ -62,6 +65,9 @@ public String getIsbn() {
     }
 
     public void setIsbn(String isbn) {
+        Preconditions.checkNotNull(isbn, String.format(ErrorConstants.NOT_NULL, "isbn"));
+        Preconditions.checkArgument(isbn.matches("^(0|[1-9][0-9]*)$\n"),
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: I recommend using this library to do the check by numbers:

import org.apache.commons.lang.StringUtils; ---> import
StringUtils.isNumeric(isbn);

File: src/main/java/wolox/training/models/Book.java [hunk 18]
```diff
@@ -78,6 +88,9 @@ public String getYear() {
     }
 
     public void setYear(String year) {
+        Preconditions.checkNotNull(year, String.format(ErrorConstants.NOT_NULL, "year"));
+        Preconditions.checkArgument(year.matches("^(0|[1-9][0-9]*)$\n"),
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: I recommend using this library to do the check by numbers:

import org.apache.commons.lang.StringUtils; ---> import
StringUtils.isNumeric(year);

File: src/main/java/wolox/training/models/User.java [hunk 19]
```diff
@@ -71,7 +71,7 @@ public LocalDate getBirthDate() {
 
     public void setBirthDate(LocalDate birthDate) {
         Preconditions.checkNotNull(birthDate, String.format(ErrorConstants.NOT_NULL, "birthDate"));
-        Preconditions.checkNotNull(LocalDate.now().compareTo(birthDate),
+        Preconditions.checkNotNull(LocalDate.now().isAfter(birthDate),
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: Change for checkArgument

**Review by Gianni-Nicolas — CHANGES_REQUESTED** [review]

**Review by Gianni-Nicolas — DISMISSED** [review]

**Review by AleFalcon — APPROVED** [review]
It's Ok. Great Job

**Review by Gianni-Nicolas — APPROVED** [review]

**Pull request merged**